torch/csrc/jit/passes/graph_fuser.cpp

```diff
@@ -141,33 +186,51 @@ struct GraphFuser {
   }
 
   bool hasSupportedType(Node* node) {
-    return areTensorsOfSameShape(node->inputs()) &&
-           haveSupportedType(node->inputs()) &&
+    return haveSupportedShape(node->inputs()) &&
+           haveSupportedType(node->inputs(), /*allow_constant_numbers*/true) &&
            haveSupportedType(node->outputs());
   }
 
-  bool haveSupportedType(at::ArrayRef<Value*> list) {
+  bool haveSupportedType(at::ArrayRef<Value*> list, bool allow_constant_numbers=false) {
     for (Value *v : list) {
-      if (!hasSupportedType(v)) return false;
+      if (!hasSupportedType(v, allow_constant_numbers)) return false;
     }
     return true;
   }
 
+  bool haveSupportedShape(at::ArrayRef<Value*> list) {
+    std::vector<Value*> tensor_values;
+    tensor_values.reserve(list.size());
+    for (Value * v : list) {
+      if (v->type()->cast<TensorType>()) {
+        tensor_values.push_back(v);
+        continue;
+      }
+      // XXX: For now, we support fusing tensor-scalar ops if the scalars
+      // are constant. They are directly inlined into the kernel body
+      if (v->type()->isSubtypeOf(NumberType::get()) &&
+          v->node()->kind() != prim::Constant) {
+        return false;
+      }
+    }
```

Comment:
I don't really like this function, because we end up with the hacky checks we had previously. If we want to handle tensor-scalar ops, this function should explicitly look for them using `->matches()`, and have a fallback that expect all-tensor inputs for all other pointwise ops. It would be much clearer too. You can leave the code in `hasSupportedType(Node*)` as it was, and put the `matches` checks before to shortcut the scalar cases. This way we wouldn't need "allow constant numbers" either - you could just slice out the tensor inputs.